packages/typescript/src/watchProgram.ts

```diff
@@ -3,20 +3,113 @@ import { PluginContext } from 'rollup';
 import { buildDiagnosticReporter } from './diagnostics/emit';
 import { DiagnosticsHost } from './diagnostics/host';
 import { Resolver } from './moduleResolution';
+import { DiagnosticCategory } from 'typescript';
 
 type BuilderProgram = import('typescript').EmitAndSemanticDiagnosticsBuilderProgram;
 
+// @see https://github.com/microsoft/TypeScript/blob/master/src/compiler/diagnosticMessages.json
+enum DiagnosticCode {
+  FILE_CHANGE_DETECTED = 6032,
+  FOUND_1_ERROR_WATCHING_FOR_FILE_CHANGES = 6193,
+  FOUND_N_ERRORS_WATCHING_FOR_FILE_CHANGES = 6194
+}
+
 interface CreateProgramOptions {
   /** Formatting host used to get some system functions and emit type errors. */
   formatHost: DiagnosticsHost;
   /** Parsed Typescript compiler options. */
   parsedOptions: import('typescript').ParsedCommandLine;
   /** Callback to save compiled files in memory. */
   writeFile: import('typescript').WriteFileCallback;
+  /** Callback for the Typescript status reporter. */
+  status: import('typescript').WatchStatusReporter;
   /** Function to resolve a module location */
   resolveModule: Resolver;
 }
 
+type DeferredResolve = ((value?: boolean) => void) | (() => void)
+
+interface Deferred {
+  promise: Promise<boolean | void>;
+  resolve: DeferredResolve;
+}
+
+function createDeferred (timeout?: number): Deferred {
+  let promise: Promise<boolean | void>
+  let resolve: DeferredResolve = () => {}
+
+  if (timeout) {
+    promise = Promise.race<Promise<boolean>>([
+      new Promise(r => setTimeout(r, timeout, true)),
+      new Promise(r => resolve = r)
+    ]);
+  } else {
+    promise = new Promise(r => resolve = r);
+  }
+
+  return { promise, resolve };
+}
+
+/**
+ * Typescript watch program helper to sync Typescript watch status with Rollup hooks.
+ */
+export class WatchProgramHelper {
+  private _startDeferred: Deferred | null = null;
+  private _finishDeferred: Deferred | null = null;
+
+  watch (timeout: number = 1000) {
```

Comment:
Please make sure to run prettier to fix the styling. (We should make sure the CI is running it later.) There shouldn't be spaces between function names and parenthesis.